### ISSUE TITLE
PSWebsocket still causing crash?

### ISSUE THREAD

**Issue by Prudwell** [couchbase/couchbase-lite-ios#1514]
I'm using the [react-native-couchbase-lite](https://github.com/couchbaselabs/react-native-couchbase-lite) package for iOS development and am getting a crash when running the app on my phone (no crash in simulator). Watching the console.logs from react native, the app crashes when it reaches a replication cancellation. 

`                database.replicate('consumer_app', remoteObject, {"cancel": true, "continuous": true})
`

I have seen issue #1369 and that led me to try the latest cblite (I was on 1.3.0 previously) but I'm still getting the crash when using 1.3.1. I've also tried building couchbase lite from the current master branch and using that I still get the crash.

<img width="298" alt="screen shot 2016-11-17 at 1 46 15 pm" src="https://cloud.githubusercontent.com/assets/11758003/20409205/982053d2-accd-11e6-885e-07598bf1d9c5.png">

I see barrier_sync in that stack, not far above PSWebsocket, and I thought the barrier_sync stuff was removed after #1369. 

The xcode log just displays '(lldb)', no verbose explanation of what is stopping the code. The error is an EXC_BREAKPOINT, so I can step forward in execution in xcode.

I'm a bit at a loss of what to try next. Are there other libraries I need to ensure are updated as well? Does this seem like potentially a different problem?

Cheers

--------
* Version:   CBLite 1.3.1
* Client OS: iOS 10.1.1
* Server:    SG 1.3.0

**Comment by pasin** [comment]
Can you do `bt all` to get the full stack trace? Log with 'Sync' and 'SyncVerbose' enabled could be really helpful?

**Comment by Prudwell** [comment]
Ok got the backtrace and the xcode logs right before the crash with Sync and SyncVerbose logging turned on.

[https://gist.github.com/Prudwell/00c6fa33e506415c1d6e58c0dcc149ce](https://gist.github.com/Prudwell/00c6fa33e506415c1d6e58c0dcc149ce)

**Comment by pasin** [comment]
I have checked both 1.3.1 and master branch, and both of them got `dispatch_barrier_sync` removing fix from the PSWebsocket library. Can you double check if CBL library got updated correctly from react-native app?

**Comment by Prudwell** [comment]
Thank you for the tip. 

This is very strange, it is indeed still showing 1.3.0 Build 61 in the log during startup for the simulator. I've cleaned and rebuilded many times, linked, deleted, re-linked the new couchbase libraries and made sure they're pointing to the correct folder location of the new 1.3.1 build, it still says it's running 1.3.0. I've tried resetting the content and settings of the simulator, then running again, still says it's running 1.3.0. I've deleted the 1.3.0 folder completely from my disk just in case somewhere xcode was still linking to that location but no luck. Xcode is doing something Xcode-ish it seems. Will keep investigating and update this if I find what's wrong. I'm guessing it's something very simple that I just don't know about yet when it comes to Xcode development.

**Comment by Prudwell** [comment]
Have tried most of the steps [here](http://stackoverflow.com/questions/5714372/how-to-empty-caches-and-clean-all-targets-xcode-4) to make sure Xcode isn't using some cached version of a previous CBL version but no luck. 

I've done a Clean Build Folder from within Xcode as well as manually deleted the _~/Library/Developer/Xcode/DerivedData_ and rebuilt + resetting simulator; no solution yet.  I'm out of ideas at the moment, will revisit tomorrow.

**Comment by pasin** [comment]
@Prudwell I'm not familiar with the react-native project. @jamiltz is the best person here to help on this.

**Comment by jamiltz** [comment]
I don't have any idea of what might be happening here. As part of the plugin installation, one step is to download and import Couchbase Lite (the plugin doesn't do it automatically). @Prudwell did you figure it out?

**Comment by Prudwell** [comment]
I took a break from this as I was tired of bashing my head against the keyboard instead of building features. 

Took another look this morning and voila! It indeed turned out to be an eccentricity of Xcode. There were old versions of the CouchbaseLite.Framework and CouchBaseLiteListener.Framework actually stored in the ios folder of the react native project. When I removed the reference from within Xcode to those libraries, it did not delete those folders, though the Xcode UI made it look like those folders were not part of the Xcode project anymore. Ok fine, working as expected since I didn't click the Move To Trash option. 

However, when I added the new CBL Framework files to Xcode from the 1.3.1 folder, it did NOT honor that new location and instead defaulted to using the versions it saw in the project folder. It gave no indication that linking to the new files failed. 

I discovered this when I tried using the "Copy Files if Needed" option when adding the Frameworks to the Xcode project this morning. It failed with a duplicate error, which confused me as I was sure I had removed the files from Xcode. 

[This comment](http://stackoverflow.com/a/26579783/4766311) led me to check the actual folder from Finder, not from within Xcode's psuedo folder browser.

If there's ever documentation on upgrading Couchbase Lite, it might be worth adding a sentence to the iOS section to help people like me avoid this incredibly frustrating problem that I wasted days on. 

Thanks for helping me out with this guys!
